aiven/resource_service.go

```diff
@@ -1010,3 +1045,175 @@ func copyConnectionInfoFromAPIResponseToTerraform(
+	return resourceServiceGetServicePlanParametersInternal(ctx, client, project, service.Type, service.Plan)
+}
+
+func resourceServiceGetServicePlanParametersFromSchema(ctx context.Context, client *aiven.Client, d resourceStateOrResourceDiff) (servicePlanParameters, error) {
+	project := d.Get("project").(string)
+	serviceType := d.Get("service_type").(string)
+	servicePlan := d.Get("plan").(string)
+
+	return resourceServiceGetServicePlanParametersInternal(ctx, client, project, serviceType, servicePlan)
+}
+
+func resourceServiceGetServicePlanParametersInternal(_ context.Context, client *aiven.Client, project, serviceType, servicePlan string) (servicePlanParameters, error) {
+	servicePlanResponse, err := client.ServiceTypes.GetPlan(project, serviceType, servicePlan)
+	if err != nil {
+		return servicePlanParameters{}, fmt.Errorf("unable to get service plan from api: %w", err)
+	}
+	return servicePlanParameters{
+		diskSizeMBDefault: servicePlanResponse.DiskSpaceMB,
+		diskSizeMBMax:     servicePlanResponse.DiskSpaceCapMB,
+		diskSizeMBStep:    servicePlanResponse.DiskSpaceStepMB,
+	}, nil
+}
+
+func resourceServiceDynamicDiskSpaceIsAllowedByPricing(_ context.Context, client *aiven.Client, d resourceStateOrResourceDiff) (bool, error) {
+	// to check if dynamic disk space is allowed, we currently have to check
+	// the pricing api to see if the `extra_disk_price_per_gb_usd` field is set
+
+	project := d.Get("project").(string)
+	serviceType := d.Get("service_type").(string)
+	servicePlan := d.Get("plan").(string)
+	cloudName := d.Get("cloud_name").(string)
+
+	servicePlanPricingResponse, err := client.ServiceTypes.GetPlanPricing(project, serviceType, servicePlan, cloudName)
+	if err != nil {
+		return false, fmt.Errorf("unable to get service plan pricing from api: %w", err)
+	}
+	return len(servicePlanPricingResponse.ExtraDiskPricePerGBUSD) > 0, nil
+}
+
+// resourceServiceCustomizeDiff validates that the diff has a chance to be applied
+func resourceServiceCustomizeDiff(ctx context.Context, d *schema.ResourceDiff, m interface{}) error {
+	// Some service integrations can be created when creating a service, think 'read_replica' for example.
+	// Those are only applicable for the initial request and should not
+	if err := resourceServiceCustomizeDiffCheckNoNewServiceIntegrationAfterCreation(ctx, d, m); err != nil {
+		return fmt.Errorf("failed service integration check: %w", err)
+	}
+	// We need to validate that the configured disk space matches the service plan; consider the situation
+	// that we have custom disk space configured but want to migrate into a cloud and plan that do not support the configured
+	// disk space anymore
+	if err := resourceServiceCustomizeDiffCheckDiskSpace(ctx, d, m); err != nil {
+		return fmt.Errorf("failed disk space check: %w", err)
+	}
+
+	return nil
+}
+
+func resourceServiceCustomizeDiffCheckNoNewServiceIntegrationAfterCreation(ctx context.Context, d *schema.ResourceDiff, m interface{}) error {
```

Comment:
looks like that we are not using a `context` here, do we need it? At least we can use a `_`